oasislmf/model_preparation/summaries.py

```diff
@@ -413,3 +419,153 @@ def generate_summaryxref_files(model_run_fp, analysis_settings):
+        tiv_sum = df.loc[
+            (df['peril_id']==peril_code) &\
+            (df['coverage_type_id']==SUPPORTED_COVERAGE_TYPES[coverage_type]['id']),
+            'tiv'].sum()
+        tiv_sum = float(tiv_sum)
+        exposure_summary[peril][status]['tiv_by_coverage'][coverage_type] = tiv_sum
+        if coverage_type in exposure_summary[peril]['all']['tiv_by_coverage']:
+            exposure_summary[peril]['all']['tiv_by_coverage'][coverage_type] += tiv_sum
+        else:
+            exposure_summary[peril]['all']['tiv_by_coverage'][coverage_type] = tiv_sum
+        exposure_summary[peril][status]['tiv'] += tiv_sum
+        exposure_summary[peril]['all']['tiv'] += tiv_sum
+
+    # Find number of locations
+    loc_count = df.loc[df['peril_id']==peril_code, loc_num].drop_duplicates().count()
+    loc_count = int(loc_count)
+    exposure_summary[peril][status]['number_of_locations'] = loc_count
+    exposure_summary[peril]['all']['number_of_locations'] += loc_count
+
+    return exposure_summary
+
+
+@oasis_log
+def write_exposure_summary(
+    target_dir,
+    gul_inputs_df,
+    exposure_fp,
+    keys_errors_fp,
+    exposure_profile,
+    loc_num
+):
+    """
+    Create exposure summary as dictionary of TIVs and number of locations
+    grouped by peril and validity respectively. Writes dictionary as json file
+    to disk.
+
+    :param target_dir: directory on disk to write exposure summary file
+    :type target_dir: str
+
+    :param gul_inputs_df: dataframe from gul_inputs.get_gul_input_items(..)
+    :type gul_inputs_df: pandas.DataFrame
+
+    :param exposure_fp: file path to input exposure file
+    :type exposure_fp: str
+
+    :param keys_errors_fp: file path to keys erors file
+    :type keys_errors_fp: str
+
+    :param exposure_profile: profile defining exposure file
+    :type exposure_profile: dict
+
+    :param loc_num: location number column heading from exposure file
+    :type loc_num: str
+    """
+
+    # Get keys in GUL input items dataframe and merge with key errors to obtain
+    # status information
+    try:
+        gul_inputs_errors_df, _ = get_gul_input_items(
+            exposure_fp, keys_errors_fp, exposure_profile=exposure_profile
+        )
+    except OasisException:   # Empty dataframe
+        gul_inputs_errors_df = pd.DataFrame(columns=gul_inputs_df.columns)
+    keys_errors_df = pd.read_csv(keys_errors_fp)
+    keys_errors_df.columns = keys_errors_df.columns.str.lower()
+    gul_inputs_errors_df = gul_inputs_errors_df.merge(
+        keys_errors_df,
+        left_on=[loc_num, 'peril_id', 'coverage_type_id'],
+        right_on=['locid', 'perilid', 'coveragetypeid'],
+        how='left'
+    )
+    gul_inputs_errors_df.drop(
+        columns=['locid', 'perilid', 'coveragetypeid', 'message'],
+        axis=1,
+        inplace=True
+    )
+
+    # Compile summary of exposure data
+    exposure_summary = {}
+    for peril_code in gul_inputs_df['peril_id'].drop_duplicates():
```

Comment:
Are you trying to loop over the set of peril IDs in the `peril_id` column or over the list of peril IDs in the column? If you want a set then you can use the `unique()` method on the column. It is best to avoid `for` loops in Pandas rows, or, if you have to use one, at least minimise the number of iterations - I think perhaps you may want `groupby` here, i.e. to group the rows in the GUL inputs dataframe by peril ID values, and then deal with each group separately.

Reply:
@sr-murthy I want to look over the set of unique peril IDs. I was under the impression that there was no discernible difference between the `.drop_duplicates()` and `.unique()` methods. The `for` loop should be looping over the result of `gul_inputs_df['peril_id'].drop_duplicates()` which returns a numpy array. The `for` loop shouldn't be looping over any rows in the dataframe.

Reply:
For the case of a single column, there's probably no difference, but I have always used `unique()` when looking at the underlying set of values in a column, and I think that would be preferable.

Reply:
OK, I understand. Actually, I was wrong: `.drop_duplicates()` returns a pandas Series, `.unique()` returns a numpy array. I've changed it to `gul_inputs_df['peril_id'].unique()`.

Reply:
OK. A Pandas series is basically a wrapper around a Numpy 1d-array, and a Pandas dataframe a wrapper around a Numpy nd-array. You can see this by using the `.values` attribute on any Pandas series or dataframe.

Reply:
This means that you can do Numpy operations on the underlying values of any Pandas series or dataframe, and often this is quicker than using Pandas operations.